### PR TITLE
Synchronize external track collections [aoide Prerequisite]

### PR DESCRIPTION
**Pull request by uklotzde**
Synchronization **from** Mixxx **to** external track collections:

- Implements only the *export* direction (one-way), i.e. no *import*. The internal track collection in Mixxx is the leader and the external track collection is the follower by receiving incremental updates.
- New interface and base class `ExternalTrackCollection`
- `Library` manages all external collections
- Changed some parameters if required, e.g. from `TrackId` to `TrackRef`
- `TrackDAO::databaseTracksMoved` has been replaced by `TrackDAO::databaseTracksReplaced` with slightly different and more useful parameters. Please don't start discussing the usage of `QPair`, that's just a negligible detail ;)
- Moved the transactions from `DirectoryDAO` into `TrackCollection`.

Considered feature complete for the first iteration.

### CODE REVIEW

**Comment by daschuer** [comment]
Did you also consider online Libraries like Spotify as ExteralTrackCollection?

File: src/library/library.cpp [hunk 1]
```diff
@@ -482,8 +499,109 @@ void Library::saveCachedTrack(Track* pTrack) noexcept {
...
+        externalTrackCollection->saveTrack(*pTrack, ExternalTrackCollection::ChangeHint::Added);
+    }
+    // Already added to m_pTrackCollection
+}
+
+void Library::slotScanTracksUpdated(QSet<TrackId> updatedTrackIds) {
+    if (updatedTrackIds.isEmpty()) {
+        return;
+    }
+    if (!m_externalTrackCollections.isEmpty()) {
```
Comment (daschuer) [inline src/library/library.cpp]: This can also become  an early exit.

**Review by daschuer — CHANGES_REQUESTED** [review]
Some comments. Thank you for the work.

File: src/library/dao/trackdao.cpp [hunk 2]
```diff
@@ -206,6 +206,37 @@ QString TrackDAO::getTrackLocation(TrackId trackId) {
...
+            "SELECT track_locations.location FROM track_locations "
+            "INNER JOIN library on library.location=track_locations.id "
+            "WHERE library.id IN (%1)";
+    {
+        QStringList idList;
+        idList.reserve(ids.size());
+        for (const auto& id : ids) {
+            idList.append(id.toString());
+        }
+        stmt = stmt.arg(idList.join(","));
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/library/dao/trackdao.cpp]: This follows the exact same pattern that is used everywhere else in this file for id lists in SQL queries. I will not change this to keep it consistent.

File: src/library/library.cpp [hunk 3]
```diff
@@ -482,8 +499,109 @@ void Library::saveCachedTrack(Track* pTrack) noexcept {
     // concurrently.
     m_pTrackCollection->exportTrackMetadata(pTrack);
 
+    if (!m_externalTrackCollections.isEmpty()) {
+        if (pTrack->getId().isValid()) {
+            if (pTrack->isDirty()) {
+                for (const auto& externalTrackCollection : m_externalTrackCollections) {
+                    externalTrackCollection->saveTrack(*pTrack, ExternalTrackCollection::ChangeHint::Modified);
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/library/library.cpp]: Only by convention. The external track collection is not supposed to touch any files that are used by Mixxx. Period.

**Comment by uklotzde** [comment]
> Did you also consider online Libraries like Spotify as ExteralTrackCollection?

`ExternalTrackCollection` only contains functions that propagate changes **from** Mixxx **to** the external collection, i.e. write-only. This is a different use case. Documentation is still missing, will follow eventually.

The interface is designed to cover only my use case and to keep the changes to a minimum. It is an adapter that serves as an *anti-corruption layer* between the new and the old world.  I'm not able to repair the internal track management that is wildly scattered among `Library`, `TrackCollection`, `TrackDAO`, and the `LibraryScanner` without clear responsibilities and layers of abstractions. The goal was to make the integration work without breaking anything.

File: src/library/trackcollection.cpp [hunk 4]
```diff
@@ -79,10 +94,17 @@ void TrackCollection::relocateDirectory(QString oldDir, QString newDir) {
...
+            m_directoryDao.relocateDirectory(oldDir, newDir);
+    transaction.commit();
 
-    m_trackDao.databaseTracksMoved(std::move(movedIds), QSet<TrackId>());
+    QList<QPair<TrackRef, TrackRef>> replacedTrackRefs;
+    replacedTrackRefs.reserve(movedTrackRefs.size());
+    for (const auto& movedTrackRef : movedTrackRefs) {
+        replacedTrackRefs += qMakePair(movedTrackRef, movedTrackRef);
+    }
+    m_trackDao.databaseTracksReplaced(std::move(replacedTrackRefs));
```
[the remark this answers is not in the document]
Comment (uklotzde) [reply src/library/trackcollection.cpp]: I agree, this needs to be reconsidered. I just adapted the code technically to keep it working.

`TrackCollection` delegates back to `TrackDAO` not for updating the database but just to dispatch some notifications. It somehow works, but I refuse to touch this brittle design.

**Comment by daschuer** [comment]
OK, Thank you. LGTM

**Pull request merged**